Nexus MVP

elipe17 opened the pull request:
## Summary of Changes
- Stood Nexus up on Raft internal cluster
- Created users/roles/repos for Nexus 
- Connected pipelines to Nexus for image building/pulling
Pull request closes #2458 

## How to Test

1. Follow the instructions for the `Nexus Admin UI login` and the `Local Docker Login`
2. Docker compose up your frontend and backend. If everything worked from step 1, you should be pulling all of your images from Nexus now.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Nexus is accessible via a service account
- [x] Nexus is automatically updated with newly built images
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any new security issues?
+ [x] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Jsut committing to not lose state

elipe17 committed:
- Updated docs
- Updated compose files to name tdp containers better

elipe17 committed:
- remove unneeded nexus docker stuff

elipe17 committed:
- adding tag command for circleci
- commenting build and test for fast iteration
- created init command to avoid tagging on dev environments

elipe17 committed:
- remove commented code

elipe17 committed:
- make executable

elipe17 committed:
- rename command

elipe17 committed:
- rename init command

elipe17 committed:
- move init back to original command
- add init to docker for now

elipe17 committed:
- fixed yaml
- re-added init command

elipe17 committed:
- password from stdin

elipe17 committed:
- Update env name

elipe17 committed:
- getting directory structure

elipe17 committed:
- tagging for frontend and backend

elipe17 committed:
- remove debug code

elipe17 committed:
- consolidate command

elipe17 committed:
- add missing step

elipe17 committed:
- Revert build and test workflow
- Add image tagging for deployments
- Update docs for admin login to Nexus UI

elipe17 committed:
- Point compose to nexus

elipe17 committed:
- Update variables on deployments

elipe17 committed:
- add docker login docs

elipe17 committed:
- title update

elipe17 committed:
- add docker login for build and test

elipe17 committed:
- fix indent

andrew-jameson commented on .circleci/deployment/jobs.yml:
```diff
@@ -133,6 +142,10 @@
     executor: docker-executor
     working_directory: ~/tdp-deploy
     steps:
+      - init-deploy
+      - build-and-tag-images:
+          backend-appname: tdp-backend-prod
+          frontend-appname: tdp-frontend-prod
```
As this is under the `deploy-production` job, will we run into issues of this running from the HHS CircleCI? Will the `CIRCI_DOCKER_LOGIN` variable would need to be recreated there?

elipe17 replied:
Yes, it will need to be added if we want all envs to have their own specific tags

andrew-jameson commented on .circleci/deployment/commands.yml:
```diff
@@ -1,4 +1,30 @@
+    steps:
+      - setup_remote_docker
+      - run:
+          name: Tag images
```
I see that we're just tagging/uploading. Beyond MVP, how should we prevent duplication of our [rafttech hub](https://hub.docker.com/repository/docker/rafttech/tdp-backend/general) like is used by [clamav here](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-backend/manifest.clamav.yml#L8)?

elipe17 replied:
Im not sure what you're asking, can you digress a bit more? The existing raftech dependency images can stay in that repo or we can move them to Nexus. They shouldnt change in our raftech repo so there isn't a dire need to move them over unless we just want everything in one place. As for the rest of the dependencies, those arent being re-tagged/uploaded but they do exist in Nexus and are being pulled from it.

elipe17 committed:
- Debugging

elipe17 committed:
- Build images for both architectures

elipe17 committed:
- add layer caching

elipe17 committed:
- stop docker caching

andrew-jameson commented on .circleci/generate_config.sh:
Is this an artifact of something gone wrong/missed? `generate_config.sh` in develop is not empty?

andrew-jameson commented on .circleci/deployment/commands.yml:
```diff
@@ -1,4 +1,39 @@
+          command: |
+            BUILD_DATE=`date +%F`
+            echo "$CIRCI_DOCKER_LOGIN" | docker login https://tdp-docker.dev.raftlabs.tech -u tdp-circi --password-stdin
+            docker build --platform linux/amd64 -t tdp-docker.dev.raftlabs.tech/<<parameters.backend-appname>>:$BUILD_DATE -t tdp-docker.dev.raftlabs.tech/<<parameters.backend-appname>>:latest ./tdrs-backend
```
Can we wrap these in a bash function for easier abstraction? I'd also like to see a combo of `$BUILD_DATE_build_$BUILD_NUM_$COMMIT_HASH` or something along those lines for truely unique identifiers although the 'latest' pointer you had is slick!

elipe17 committed:
Merge branch 'develop' into 2458-nexus-mvp

elipe17 committed:
- add vm access instructions

elipe17 committed:
- Added script for image building/tagging
- Using script in pipeline

elipe17 committed:
- testing

elipe17 committed:
- fix path

elipe17 committed:
- use workdir var
- debug statements

elipe17 committed:
- quoting paths
- extra debugging

elipe17 committed:
- debug

elipe17 committed:
- passing current dir
- still debugging

elipe17 committed:
- remove debugging code
- always let script keep running

elipe17 committed:
- adding step for docker buildx

elipe17 committed:
- comment to increase speed

elipe17 committed:
- install dep
- check docker version

elipe17 committed:
- add containerd

elipe17 committed:
- trying larger resource class
- tag update

elipe17 committed:
- remove hash

elipe17 committed:
- quiet

elipe17 committed:
- leave output because circi requires it

elipe17 committed:
- test dockerd max uploads

elipe17 committed:
- try restarting daemon with max uploads

elipe17 committed:
- test individual executors
- comment out buildx for now
- add build and tag jobs
- update workflow
- update script

elipe17 committed:
- deploy after build

elipe17 committed:
- depend on infra

elipe17 committed:
- explicit push

elipe17 committed:
- one build executor at a time

elipe17 committed:
- build after deploy

elipe17 committed:
- try arm medium

elipe17 committed:
- revert back to OG working build state

elipe17 committed:
- Test using non docker in docker

elipe17 committed:
- Update docker daemon config
- Create new builder

elipe17 committed:
- Don't do entire init-deploy routine

elipe17 committed:
- Don't push from build ctx since it doesnt obey daemon config single layer push

elipe17 committed:
- Update OS image
- Remove resource class

elipe17 committed:
- Remove build and tag from dev envs
- Update develop, staging, and prod to build and tag before deploying

elipe17 committed:
- remove extra inclusion of "branch" tag

elipe17 committed:
- remove unneeded variable
- add a job for each env

reitermb commented:
On deck for office hours today

elipe17 commented on .circleci/build-and-test/jobs.yml:
```diff
@@ -3,11 +3,13 @@
           name: Run Unit Tests And Create Code Coverage Report
           command: |
             cd tdrs-backend;
+            docker ps
```
remove that

ADPennington replied:
before merge @elipe17 ?

elipe17 committed:
- add docker login for owasp scan

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2458-nexus-mvp

elipe17 committed:
Merge branch 'develop' into 2458-nexus-mvp

ADPennington commented:
@vlasse86 -- @elipe17 and I will pair on this ticket for qasp review on 10/2.

elipe17 committed:
Merge branch 'develop' into 2458-nexus-mvp

ADPennington commented on .circleci/build-and-test/commands.yml:
```diff
@@ -49,6 +49,13 @@
+      - run:
+          name: Docker login
+          command: |
+            echo "$CIRCI_DOCKER_LOGIN" | docker login https://tdp-docker.dev.raftlabs.tech -u tdp-circi --password-stdin
```
@ADPennington added this env var to HHS circleci. cc @elipe17

ADPennington reviewed (approved):
LGTM @elipe17 🚀 

@elipe17 this can be merged as soon as you revert the one change.

ADPennington committed:
Merge branch 'develop' into 2458-nexus-mvp

elipe17 committed:
Merge branch 'develop' into 2458-nexus-mvp

elipe17 committed:
- remove debug cmd

elipe17 committed:
Merge branch '2458-nexus-mvp' of https://github.com/raft-tech/TANF-app into 2458-nexus-mvp
